### PR TITLE
Expose a LoggingDnsQueryLifeCycleObserverFactory

### LINKED ISSUE #10485: Ability to change the logger and level used by DnsNameResolverBuilder#traceEnabled
### Expected behavior
Ability to change the logger and level used by `DnsNameResolverBuilder#traceEnabled`

### Actual behavior
When `DnsNameResolverBuilder#traceEnabled` is enabled, it uses always the logger and level below
https://github.com/netty/netty/blob/46cb4015ff25ff2dd3d0dede5c3185ac498e56b0/resolver-dns/src/main/java/io/netty/resolver/dns/TraceDnsQueryLifeCycleObserverFactory.java#L26-L28

It would be nice if those can be configured.

Possible solutions:
- `DnsNameResolverBuilder#traceEnabled` where one can specify logger and level

Or

- `TraceDnsQueryLifeCycleObserverFactory` can be made public and also it can expose construction similar to what `LoggingHandler` provides i.e.

```
    public TraceDnsQueryLifeCycleObserverFactory(Class<?> clazz, LogLevel level) {
        ObjectUtil.checkNotNull(clazz, "clazz");
        ObjectUtil.checkNotNull(level, "level");
        this.logger = InternalLoggerFactory.getInstance(clazz);
        this.level = level.toInternalLevel();
    }
```

With this approach one will be able to add such observer with `DnsNameResolverBuilder#dnsQueryLifecycleObserverFactory`

this second approach will require a dependency to `netty-handler` module

Wdyt?

### Netty version
current snapshot

### JVM version (e.g. `java -version`)
jdk 8

### OS version (e.g. `uname -a`)
Mac OS

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
 There is a use case for having logging in the DnsNameResolver, similar to the LoggingHandler.
 Previously, one could set `traceEnabled` on the DnsNameResolverBuilder, but this is not very configurable.
 Specifically, the log level cannot be changed.

Modification:
 Expose a LoggingDnsQueryLifeCycleObserverFactory, that permit changing the log-level.

Result:
 It is now possible to get logging in the DnsNameResolver at a custom log level, without very much effort.

Fixes #10485

### CODE REVIEW

**Comment by violetagg** [comment]
@chrisvest Is it possible to have this in `4.1` where `traceEnabled` is deprecated instead of removed?

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    public LoggingDnsQueryLifeCycleObserverFactory() {
+        this(LogLevel.DEBUG);
+    }
+
+    /**
+     * Create {@link DnsQueryLifecycleObserver} instances that log events at the given log level.
+     * @param level The log level to use for logging resolver events.
+     */
+    public LoggingDnsQueryLifeCycleObserverFactory(LogLevel level) {
+        logger = InternalLoggerFactory.getInstance(getClass());
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:49]: Should we use a `static` logger and so use `InternalLoggerFactory.getInstance(LoggingDnsQueryLifeCycleObserverFactory.class)` ?

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link DnsQueryLifecycleObserverFactory} that enables detailed logging in the {@link DnsNameResolver}.
+ * <p>
+ * When {@linkplain DnsNameResolverBuilder#dnsQueryLifecycleObserverFactory(DnsQueryLifecycleObserverFactory)
+ * configured on the resolver}, detailed trace information will be generated so that it is easier to understand the
+ * cause of resolution failure.
+ */
+public class LoggingDnsQueryLifeCycleObserverFactory implements DnsQueryLifecycleObserverFactory {
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:33]: final

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2017 The Netty Project
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:2]: 2020

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by normanmaurer** [comment]
@chrisvest I think @violetagg is right and we should do this for 4.1 as well with `@deprecation`... That said for master this one is fine.

**Comment by normanmaurer** [comment]
@netty-bot whitelist

**Comment by normanmaurer** [comment]
@netty-bot add to whitelist

**Comment by chrisvest** [comment]
@normanmaurer PR comments addressed, please take a look.

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 4]
```diff
@@ -30,8 +30,9 @@
  * configured on the resolver}, detailed trace information will be generated so that it is easier to understand the
  * cause of resolution failure.
  */
-public class LoggingDnsQueryLifeCycleObserverFactory implements DnsQueryLifecycleObserverFactory {
-    private final InternalLogger logger;
+public final class LoggingDnsQueryLifeCycleObserverFactory implements DnsQueryLifecycleObserverFactory {
+    public static final InternalLogger LOGGER =
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:34]: private ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java (hunk 4 above)
Comment (chrisvest) [reply resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java]: img

**Comment by violetagg** [comment]
@chrisvest Why only the log level can be specified?
I expected to have the ability as in `LoggingHandler`.
Let's say you create some application on top of Netty and you want to have a specific logger for that specific functionality.
Typically in applications you have categories for different type of issues and enabling the logging for a particular category you can debug this specific issue.
With the current implementation we still use Netty specific logger.

**Comment by chrisvest** [comment]
@violetagg Because Netty does not have any external logging API. The `InternalLogger` interface is not meant to be accessible outside of Netty itself. `LoggingHandler` is the same in that regard. Or are you referring to the name of the logger?

**Comment by violetagg** [comment]
Is it possible something like this

```
    public LoggingDnsQueryLifeCycleObserverFactory(Class<?> clazz, LogLevel level) {
         requireNonNull(clazz, "clazz");
         this.logger = InternalLoggerFactory.getInstance(clazz);
         this.level = requireNonNull(level, "level").toInternalLevel();
     }
```

**Comment by normanmaurer** [comment]
@violetagg good call... that said use `String name` @chrisvest, just as we do in `LoggingHandler`

**Comment by chrisvest** [comment]
@normanmaurer `LoggingHandler` does both.

**Comment by normanmaurer** [comment]
lets do both then here as well

**Comment by chrisvest** [comment]
Added those constructors as well, and squashed the commits.

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * Create {@link DnsQueryLifecycleObserver} instances that log events to a logger with the given class context,
+     * at the given log level.
+     * @param classContext The class context for the logger to use.
+     * @param level The log level to use for logging resolver events.
+     */
+    public LoggingDnsQueryLifeCycleObserverFactory(Class<?> classContext, LogLevel level) {
+        this.level = checkAndConvertLevel(level);
+        logger = InternalLoggerFactory.getInstance(classContext);
```
Comment (violetagg) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:63]: May be add `requireNonNull(classContext)`

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * Create {@link DnsQueryLifecycleObserver} instances that log events to a logger with the given name context,
+     * at the given log level.
+     * @param nameContext The name context for the logger to use.
+     * @param level The log level to use for logging resolver events.
+     */
+    public LoggingDnsQueryLifeCycleObserverFactory(String nameContext, LogLevel level) {
+        this.level = checkAndConvertLevel(level);
+        logger = InternalLoggerFactory.getInstance(nameContext);
```
Comment (violetagg) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:74]: May be add `requireNonNull(nameContext)`

File: resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        logger = InternalLoggerFactory.getInstance(classContext);
+    }
+
+    /**
+     * Create {@link DnsQueryLifecycleObserver} instances that log events to a logger with the given name context,
+     * at the given log level.
+     * @param nameContext The name context for the logger to use.
+     * @param level The log level to use for logging resolver events.
+     */
+    public LoggingDnsQueryLifeCycleObserverFactory(String nameContext, LogLevel level) {
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/LoggingDnsQueryLifeCycleObserverFactory.java:72]: nit: I would just call this `name`

**Review by normanmaurer — CHANGES_REQUESTED** [review]
almost there

**Review by normanmaurer — APPROVED** [review]
Great job... thanks a lot

**Comment by normanmaurer** [comment]
@violetagg any concerns ?

**Comment by violetagg** [comment]
LGTM
thanks a lot

**Comment by normanmaurer** [comment]
@chrisvest feel free to merge once CI completes

**Pull request merged**